### ISSUE TITLE
Spurious `rustdoc::broken_intra_doc_links` reported by linking to macro defined later

### ISSUE THREAD

**Issue by unageek** [rust-lang/rust#98804]
I tried this code:

```rust
/// [`bar`]
#[macro_export]
macro_rules! foo {
    () => {};
}

/// [`foo`]
#[macro_export]
macro_rules! bar {
    () => {};
}
```

I expected to see this happen: No warning is reported by `cargo doc`.

Instead, this happened: The following warning is reported:

```
warning: unresolved link to `bar`
 --> src/issue.rs:1:7
  |
1 | /// [`bar`]
  |       ^^^ no item named `bar` in scope
  |
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
  = note: `macro_rules` named `bar` exists in this crate, but it is not in scope at this link's location
```

Despite the warning, the generated link does work as expected.

This does not happen if I change ``[`bar`]`` to ``[`crate::bar`]``.

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (46b8c23f3 2022-07-01)
binary: rustc
commit-hash: 46b8c23f3eb5e4d0e0aa27eb3f20d5b8fc3ed51f
commit-date: 2022-07-01
host: aarch64-apple-darwin
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by petrochenkov** [comment]
I suspect that the code you posted is not in the root module, right?
In that case `bar` is indeed not in scope at the location of the ``[`bar`]`` link, it's an intentional change made in https://github.com/rust-lang/rust/pull/96676.

>Despite the warning, the generated link does work as expected.

This is a compatibility measure to keep stuff on docs.rs working if possible.

**Comment by unageek** [comment]
> In that case bar is indeed not in scope at the location of the ``[`bar`]`` link

Aha! You are right! Thank you for your reply!
